### ISSUE TITLE
Thoughts: Trying to use Chauffeur as a component builder

### ISSUE THREAD

**Issue by TFAstudio** [aaronpowell/Chauffeur#76]
Hello

I have written up a way of using NuGet packages to take various components that might need to be delivered.

In NuGet packages, one should be able to 'uninstall' components as easily as install.  This means that there may or should be a way to uninstall any documenttypes and templates from the system.  At the moment, I may be able to get away with this, but at the very least I need to remove the 'signature' for the previously installed document types to keep testing - where would I find this, and is there a way to use chauffeur to remove components as well as add them?

Thanks

**Comment by aaronpowell** [comment]
Umbraco packages (which is what Chauffeur uses under the hood) doesn't have the concept of a destructive uninstall, only a non-destructive one. This means when you uninstall something you don't really get rid of the stuff that was added to the CMS like doc types.

If it was trying to solve that problem then you need to ask the question of what happens to the content created from those doc types? What happens if you had a doc type that was then used in composition? etc.. So you can see where this becomes really tricky to do.

I do provide a way to do it from the cli, using the `content-type` deliverable: https://aaronpowell.github.io/Chauffeur/deliverable-content-type.html

**Comment by TFAstudio** [comment]
ok thats a great start - but really this is more for the testing cycle...  once I have installed the components via the chauffeur delivery mechanism, it will not do it again - even if I delete them from inside Umbraco.  so where would I 'force' the delivery system to forget what has been installed without having to recreate a new site each time?

Thanks again for your help

**Comment by aaronpowell** [comment]
If you're meaning in the context of an integration test, then I have a dedicated package for doing that: https://aaronpowell.github.io/Chauffeur/testing-tools.html

But instead if you're talking about how to handle a static environment that you're manually testing against, going into the backoffice and changing things, etc. it is a little harder. Once you go into the backoffice and change stuff Chauffeur's workflow starts falling over.

Part of it could be addressed in issue #18, allowing you to re-run a previously run deliverable.

**Comment by aaronpowell** [comment]
I _think_ I'm following you, you're concern is that one will override the other? Is that correct?

Assuming it is, then it's actually a non-issue as the packaging engine in Umbraco (which Chauffeur uses) only does non-distructive imports, so when something imports it will only update, not remove, at least from memory.

If I'm missing the mark, do you want to create a demo repo to show me what you're thinking?

**Comment by TFAstudio** [comment]
Yes, the main issues are the fact that it will be unknown which other 'modules' will be installed.  Rather than maintaining XML for the DocumentType that allows these new ones, I have decided to try something along the lines of the following:

```c#
    [DeliverableName("AddDocumentTypeAllowed")]
    public class AddDocumentTypeAllowedDeliverable : Deliverable
    {
        private readonly IContentTypeService contentTypeService;
        public AddDocumentTypeAllowedDeliverable(TextReader reader, TextWriter writer, IContentTypeService contentTypeService) : base(reader, writer)
        {
            this.contentTypeService = contentTypeService;
        }

        public async override Task<DeliverableResponse> Run(string command, string[] args)
        {
            var contentTypes = contentTypeService.GetAllContentTypes();

            foreach (var contentType in contentTypes)
            {
                if (contentType.Alias != "DocTypeToAdd") continue;
                var alreadyHasType = false;
                List<ContentTypeSort> allowedTypes = new List<ContentTypeSort>();
                foreach (var allowedType in contentType.AllowedContentTypes)
                {
                    if (!alreadyHasType)
                        alreadyHasType= (allowedType.Alias == "newDocType");
                    allowedTypes.Add(allowedType);

                }
                // add the last one
                if (!alreadyHasType)
                {
                    var newSort = new ContentTypeSort(allowedTypes.Count, 0);
                    newSort.Alias = "newDocType";
                    allowedTypes.Add(newSort);

                    contentType.AllowedContentTypes = allowedTypes;

                    contentTypeService.Save(contentType);
                    await Out.WriteLineAsync($"Content Type named {contentType.Name} with an alias of {contentType.Alias} was updated");
                }

                
                break;
            }

            return DeliverableResponse.Continue;
        }
    }
```

**Comment by aaronpowell** [comment]
Wouldn't the only reason you don't know what other modules are installed be because you're doing stuff outside of Chauffeur? If you're running a mixed mode (some Chauffeur, some non-Chauffeur) then yeah, things are harder to track.

Deliveries aren't really intended to be run ad-hoc, they are meant to be a cumulative thing, so when you're creating one you know what's previously been done so you don't run those steps again. That's why you shouldn't have `install` as the first line of every single Delivery, you **know** you've already installed, so don't do it again.

**Comment by TFAstudio** [comment]
Ok, after some time away, we are back onto this process.

The reason why this looks a little strange is because we are trying to use Chauffeur 'outside the box' as a system to merge different aspects of a modular delivery system.

For example, we may have a basic website (`Basic`) that someone comes along and asks for a News module (`News`).  We have already deployed `Basic` to the site, and rather than going in and adding the various `DocumentTypes`, `Templates`, `Structure` dependencies, etc, that we would do with any client that asked for a `News` module, we would like to use Chauffeur to merge the `News` into the existing `Basic`.

The above `CustomDeliverable` should add the structure to the existing `DocumentTypes` of the `Basic` module.

We would prefer to have a `*.delivery` per module, which would taken in a partial XML `News.xml` to add to the existing `Basic`.  This way if `Basic.xml` changes, we do not have to merge that into the `News.xml` file (and any other subsequent sub-modules that may be created).

However, we have run into another issue that hopefully you can assist us with.

We have a hierarchy for our `DocumentTypes` to allow the tabs of the parent to propagate through to the children. We would like to be able to add a `DocumentType` to one of the children inside of this hierarchy without having to reference all the parent `DocumentTypes` as a path.  When we try to deliver a non-Basic module (in other words just `News.xml`), we get an error 

```
Error running the current deliverable 'package'
System.IndexOutOfRangeException: The alias 'Umbraco.Core.TopologicalSorter+DependencyField`1[System.Xml.Linq.XElement]' 
has an invalid dependency. The dependency 'CustomDocumentType' does not exist in the list of aliases
```

I thought ok, so it is not possible to add without at least referring to the parent `DocumentTypes`, but even adding those in, it looks like you need the entire set of information from the node to be inserted all the way up to the `Root`, however without giving complete information, this breaks too.

Is there a way to mitigate this?  I know this is completely different way of using this than you had intended, but I hope that my explanation of using it as a modular deployment tool makes enough sense to determine if this can be done at all.  Keep in mind, there may be other modules, such as `Personnel` or `Calendar` that may be added - in any order, with or without the other module `News`.  The only requirement is `Basic`.

Thanks again for your time and thoughts.
